bats/fb-test-puppet.bats

```diff
@@ -10,6 +10,14 @@ if [[ -e /etc/profile.d/puppet-agent.sh ]] ; then
   . /etc/profile.d/puppet-agent.sh
 fi
 
+@test "enable puppet feature" {
+  KATELLO_VERSION=$(tKatelloVersion)
+  if [[ $KATELLO_VERSION != 4.[3-9]* ]]; then
+    skip "Enabling Puppet explicitly is only needed with Katello 4.3+"
+  fi
+  foreman-installer --enable-foreman-plugin-puppet --enable-foreman-cli-puppet --foreman-proxy-puppet true --foreman-proxy-puppetca true --foreman-proxy-content-puppet true --enable-puppet --puppet-server true --puppet-server-foreman-ssl-ca /etc/pki/katello/puppet/puppet_client_ca.crt --puppet-server-foreman-ssl-cert /etc/pki/katello/puppet/puppet_client.crt --puppet-server-foreman-ssl-key /etc/pki/katello/puppet/puppet_client.key
```

Comment:
The `--enable-puppet` is required or else we run into the following:

```
 ✗ enable puppet feature
   (in test file fb-test-puppet.bats, line 18)
     `foreman-installer --enable-foreman-plugin-puppet --enable-foreman-cli-puppet --foreman-proxy-puppet true --foreman-proxy-puppetca true --foreman-proxy-content-puppet true --puppet-server true --puppet-server-foreman-ssl-ca /etc/pki/katello/puppet/puppet_client_ca.crt --puppet-server-foreman-ssl-cert /etc/pki/katello/puppet/puppet_client.crt --puppet-server-foreman-ssl-key /etc/pki/katello/puppet/puppet_client.key' failed with status 6
   2021-10-31 18:37:42 [NOTICE] [root] Loading installer configuration. This will take some time.
   2021-10-31 18:37:46 [NOTICE] [root] Running installer with log based terminal output at level NOTICE.
   2021-10-31 18:37:46 [NOTICE] [root] Use -l to set the terminal output log level to ERROR, WARN, NOTICE, INFO, or DEBUG. See --full-help for definitions.
   2021-10-31 18:37:55 [NOTICE] [configure] Starting system configuration.
   2021-10-31 18:38:03 [NOTICE] [configure] 250 configuration steps out of 1794 steps complete.
   2021-10-31 18:38:04 [ERROR ] [configure] Could not set groups on user[foreman-proxy]: Execution of '/sbin/usermod -G puppet foreman-proxy' returned 6: usermod: group 'puppet' does not exist
   2021-10-31 18:38:04 [ERROR ] [configure] /Stage[main]/Foreman_proxy::Config/User[foreman-proxy]/groups: change from  to 'puppet' failed: Could not set groups on user[foreman-proxy]: Execution of '/sbin/usermod -G puppet foreman-proxy' returned 6: usermod: group 'puppet' does not exist
   2021-10-31 18:38:06 [NOTICE] [configure] 500 configuration steps out of 1798 steps complete.
   2021-10-31 18:38:06 [NOTICE] [configure] 750 configuration steps out of 1800 steps complete.
   2021-10-31 18:38:08 [NOTICE] [configure] 1000 configuration steps out of 1806 steps complete.
   2021-10-31 18:38:09 [NOTICE] [configure] 1250 configuration steps out of 1809 steps complete.
   2021-10-31 18:38:25 [NOTICE] [configure] 1500 configuration steps out of 1810 steps complete.
   2021-10-31 18:38:30 [NOTICE] [configure] 1750 configuration steps out of 1810 steps complete.
```

Reply:
That's weird and unexpected.

What's I've done in the past is to remove cleanups in Kafo (comment out https://github.com/theforeman/kafo/blob/c184f41394f43c975d9152cbcf6f7fe28366a7a1/lib/kafo/configuration.rb#L189). Then inspect the files it generated in the execution environment. That can help you understand what's going on (perhaps we need an env var so you don't need to patch the source). I would be interested in the diff between an ExecEnv with `--enable-puppet` and without. I would expect them to both generate the exact same ExecEnv.

My theory is that we have [this block here in puppet-foreman_proxy](https://github.com/theforeman/puppet-foreman_proxy/blob/3ae59e862a32045e6f5d63184cbac870fa64b6c7/manifests/config.pp#L5-L16) that's compile order dependent (`defined()` relies on compile order). There is also an equivalent [here in puppet-puppet](https://github.com/theforeman/puppet-puppet/blob/01a716ccdca8975294001ec650a4ec4144881376/manifests/server/install.pp#L5-L12). Perhaps that's wrong and doesn't end up in an actual relationship anymore?

Another thing to keep in mind is that [the compile order](https://github.com/theforeman/foreman-installer/blob/926628891b02f7408cd59ce7c0a832c94709f3b4/config/katello.yaml#L20-L27) says that `foreman_proxy` should be included before `puppet`.

Another (and this is probably cleaner) is that we can explicitly manage the puppet group as well in `puppet::server::install`. Then the `user` resource should autorequire the group and build the relationship that way.

Similarly, we should also look at automatically requiring the SSL files instead of being this compile order dependent, but I'd consider that out of scope for now.

Reply:
The diff is down below and about what I expected. Without `--enable-puppet` the puppet class never gets enabled, and Kafo ignores it thus any parameters on any sub-class of the module is ignored. I wonder if that implies, if we should be splitting puppet from puppet server in the answers file (if that even works)? e.g.

```
puppet: false
puppet::server: false
```

Otherwise, I think we just have to stick with having `--enable-puppet`.

```
< puppet::version: present
< puppet::user: puppet
< puppet::group: puppet
< puppet::dir: "/etc/puppetlabs/puppet"
< puppet::codedir: "/etc/puppetlabs/code"
< puppet::vardir: "/opt/puppetlabs/puppet/cache"
< puppet::logdir: "/var/log/puppetlabs/puppet"
< puppet::rundir: "/var/run/puppetlabs"
< puppet::ssldir: "/etc/puppetlabs/puppet/ssl"
< puppet::sharedir: "/opt/puppetlabs/puppet"
< puppet::manage_packages: true
< puppet::dir_owner: root
< puppet::dir_group: 
< puppet::package_provider: 
< puppet::package_install_options: 
< puppet::package_source: 
< puppet::port: 8140
< puppet::pluginsync: true
< puppet::splay: false
< puppet::splaylimit: 1800
< puppet::autosign: "/etc/puppetlabs/puppet/autosign.conf"
< puppet::autosign_entries: []
< puppet::autosign_mode: '0664'
< puppet::autosign_content: 
< puppet::autosign_source: 
< puppet::runinterval: 1800
< puppet::usecacheonfailure: true
< puppet::runmode: service
< puppet::run_hour: 
< puppet::run_minute: 
< puppet::unavailable_runmodes: []
< puppet::cron_cmd: 
< puppet::systemd_cmd: 
< puppet::systemd_randomizeddelaysec: 0
< puppet::agent_noop: false
< puppet::show_diff: false
< puppet::module_repository: 
< puppet::http_connect_timeout: 
< puppet::http_read_timeout: 
< puppet::ca_server: 
< puppet::ca_port: 
< puppet::ca_crl_filepath: 
< puppet::prerun_command: 
< puppet::postrun_command: 
< puppet::dns_alt_names: []
< puppet::use_srv_records: false
< puppet::srv_domain: wareagle.example.com
< puppet::pluginsource: puppet:///plugins
< puppet::pluginfactsource: puppet:///pluginfacts
< puppet::additional_settings: {}
< puppet::agent_additional_settings: {}
< puppet::agent_restart_command: "/usr/bin/systemctl reload-or-restart puppet"
< puppet::classfile: "$statedir/classes.txt"
< puppet::hiera_config: "$confdir/hiera.yaml"
< puppet::auth_template: puppet/auth.conf.erb
< puppet::allow_any_crl_auth: false
< puppet::auth_allowed:
< - "$1"
< puppet::client_package:
< - puppet-agent
< puppet::agent: true
< puppet::report: true
< puppet::client_certname: pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::puppetmaster: 
< puppet::systemd_unit_name: puppet-run
< puppet::service_name: puppet
< puppet::syslogfacility: 
< puppet::environment: production
< puppet::server: true
< puppet::server_admin_api_whitelist:
< - localhost
< - pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::server_manage_user: true
< puppet::server_user: puppet
< puppet::server_group: puppet
< puppet::server_dir: "/etc/puppetlabs/puppet"
< puppet::server_ip: 0.0.0.0
< puppet::server_port: 8140
< puppet::server_ca: true
< puppet::server_ca_crl_sync: false
< puppet::server_crl_enable: 
< puppet::server_ca_auth_required: true
< puppet::server_ca_client_self_delete: false
< puppet::server_ca_client_whitelist:
< - localhost
< - pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::server_custom_trusted_oid_mapping: 
< puppet::server_http: false
< puppet::server_http_port: 8139
< puppet::server_reports: foreman
< puppet::server_puppetserver_dir: "/etc/puppetlabs/puppetserver"
< puppet::server_puppetserver_vardir: "/opt/puppetlabs/server/data/puppetserver"
< puppet::server_puppetserver_rundir: "/var/run/puppetlabs/puppetserver"
< puppet::server_puppetserver_logdir: "/var/log/puppetlabs/puppetserver"
< puppet::server_puppetserver_version: 
< puppet::server_external_nodes: "/etc/puppetlabs/puppet/node.rb"
< puppet::server_trusted_external_command: 
< puppet::server_cipher_suites:
< - TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
< - TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
< - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
< - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
< - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
< - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
< puppet::server_connect_timeout: 120000
< puppet::server_git_repo: false
< puppet::server_default_manifest: false
< puppet::server_default_manifest_path: "/etc/puppet/manifests/default_manifest.pp"
< puppet::server_default_manifest_content: ''
< puppet::server_environments_owner: puppet
< puppet::server_environments_group: 
< puppet::server_environments_mode: '0755'
< puppet::server_envs_dir: "/etc/puppetlabs/code/environments"
< puppet::server_envs_target: 
< puppet::server_common_modules_path:
< - "/etc/puppetlabs/code/environments/common"
< - "/etc/puppetlabs/code/modules"
< - "/opt/puppetlabs/puppet/modules"
< - "/usr/share/puppet/modules"
< puppet::server_git_repo_mode: '0755'
< puppet::server_git_repo_path: "/opt/puppetlabs/puppet/cache/puppet.git"
< puppet::server_git_repo_group: puppet
< puppet::server_git_repo_user: puppet
< puppet::server_git_branch_map: {}
< puppet::server_idle_timeout: 1200000
< puppet::server_post_hook_content: puppet/server/post-receive.erb
< puppet::server_post_hook_name: post-receive
< puppet::server_storeconfigs: false
< puppet::server_ruby_load_paths:
< - "/opt/puppetlabs/puppet/lib/ruby/vendor_ruby"
< puppet::server_ssl_dir: "/etc/puppetlabs/puppet/ssl"
< puppet::server_ssl_dir_manage: true
< puppet::server_ssl_key_manage: true
< puppet::server_ssl_protocols:
< - TLSv1.2
< puppet::server_ssl_chain_filepath: 
< puppet::server_package: 
< puppet::server_version: 
< puppet::server_certname: pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::server_request_timeout: 60
< puppet::server_strict_variables: false
< puppet::server_additional_settings: {}
< puppet::server_foreman: true
< puppet::server_foreman_url: https://pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::server_foreman_ssl_ca: "/etc/pki/katello/puppet/puppet_client_ca.crt"
< puppet::server_foreman_ssl_cert: "/etc/pki/katello/puppet/puppet_client.crt"
< puppet::server_foreman_ssl_key: "/etc/pki/katello/puppet/puppet_client.key"
< puppet::server_foreman_facts: true
< puppet::server_puppet_basedir: "/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet"
< puppet::server_parser: current
< puppet::server_environment_timeout: 
< puppet::server_jvm_java_bin: "/usr/bin/java"
< puppet::server_jvm_config: "/etc/sysconfig/puppetserver"
< puppet::server_jvm_min_heap_size: 2G
< puppet::server_jvm_max_heap_size: 2G
< puppet::server_jvm_extra_args: 
< puppet::server_jvm_cli_args: 
< puppet::server_jruby_gem_home: "/opt/puppetlabs/server/data/puppetserver/jruby-gems"
< puppet::server_max_active_instances: 2
< puppet::server_max_requests_per_instance: 0
< puppet::server_max_queued_requests: 0
< puppet::server_max_retry_delay: 1800
< puppet::server_multithreaded: false
< puppet::server_use_legacy_auth_conf: false
< puppet::server_check_for_updates: true
< puppet::server_environment_class_cache_enabled: false
< puppet::server_allow_header_cert_info: false
< puppet::server_web_idle_timeout: 30000
< puppet::server_puppetserver_jruby9k: false
< puppet::server_puppetserver_metrics: false
< puppet::server_puppetserver_profiler: false
< puppet::server_metrics_jmx_enable: true
< puppet::server_metrics_graphite_enable: false
< puppet::server_metrics_graphite_host: 127.0.0.1
< puppet::server_metrics_graphite_port: 2003
< puppet::server_metrics_server_id: pipe-katello-server-nightly-centos7.wareagle.example.com
< puppet::server_metrics_graphite_interval: 5
< puppet::server_metrics_allowed: 
< puppet::server_puppetserver_experimental: true
< puppet::server_puppetserver_auth_template: 
< puppet::server_puppetserver_trusted_agents: []
< puppet::server_puppetserver_trusted_certificate_extensions: []
< puppet::server_compile_mode: 
< puppet::server_acceptor_threads: 
< puppet::server_selector_threads: 
< puppet::server_ssl_acceptor_threads: 
< puppet::server_ssl_selector_threads: 
< puppet::server_max_threads: 
< puppet::server_ca_allow_sans: false
< puppet::server_ca_allow_auth_extensions: false
< puppet::server_ca_enable_infra_crl: false
< puppet::server_max_open_files: 
< puppet::server_versioned_code_id: 
< puppet::server_versioned_code_content: 
```

Reply:
> The diff is down below and about what I expected. Without `--enable-puppet` the puppet class never gets enabled, and Kafo ignores it thus any parameters on any sub-class of the module is ignored. I wonder if that implies, if we should be splitting puppet from puppet server in the answers file (if that even works)? e.g.

So the problem is that my understanding how Kafo should work and how it actually works is different.

I would be very interested to know if this is a regression we introduced at some point. If not, why my interpretation was this way. It could be that I assumed it would do this based on the answer file structure.

It also raises the question what the right behavior is. IMHO silently accepting arguments for something that's disabled is wrong. I would expect it to either enable the module or fail with a clear error that it needs to be enabled. The latter you (as a user) can solve by enabling the module or realize you made a mistake and didn't intend to do. Depending on how often users realize they made a mistake, it may just cause frustration for users.

I think for now we should include `--enable-puppet` and open an issue against Kafo to improve the handling. Do you agree?

Reply:
> I wonder if that implies, if we should be splitting puppet from puppet server in the answers file (if that even works)? e.g.

I missed this part in my reply, but splitting isn't the best idea. Technically we do have a `puppet::server` class (which it would refer to in the answers file, not `$puppet::server` in the `server` class as hiera would). However, that class is not meant to be used directly. The installer should only expose the `puppet` class.

Reply:
I don't recall a time when we hid options if the class was disbaled (which sounds like a reasonable thing for Kafo to do). Right now you see for example:

```
    --[no-]enable-puppet          Enable 'puppet' puppet module (default: false)
```

```
= Module puppet:
    --puppet-ca-server            Use a different ca server. Should be either
                                  a string with the location of the ca_server
                                  or 'false'. (current: UNDEF)
    --puppet-show-diff            Show and report changed files with diff output (current: false)
```

Which, I can see how help makes sense as you may want to include options when enabling the class. As for treating the options as invalid, I think that because it builds the set of options as above where they are available to the user, we've not included a guard later on in the code that stops them if the top level class is not enabled. I'll file a Redmine for that functionality and update this PR with the `--enable-puppet`.

Reply:
Redmine issue -- https://projects.theforeman.org/issues/33821

Reply:
> I don't recall a time when we hid options if the class was disbaled (which sounds like a reasonable thing for Kafo to do)

As you say, finding out how to enable a module and set options at the same time is IMHO a good reason to not hide modules even if they're disabled.

We also generate the man page (`man foreman-installer`) based on it and that would make the man page less usable. (sidenote: we don't generate man pages for other scenarios but we should)

> Redmine issue -- https://projects.theforeman.org/issues/33821

Thanks, I've expanded it with some context.